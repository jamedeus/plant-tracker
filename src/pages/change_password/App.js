import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword1, setNewPassword1] = useState('');
    const [newPassword2, setNewPassword2] = useState('');
    const [oldPasswordIncorrect, setOldPasswordIncorrect] = useState(false);
    const [newPasswordError, setNewPasswordError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post old and new password to backend
        const response = await fetch("/accounts/change_password/submit/", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-CSRFToken": Cookies.get('csrftoken')
            },
            credentials: "include",
            body: new URLSearchParams({
                old_password: oldPassword,
                new_password1: newPassword1,
                new_password2: newPassword2,
            }),
        });
        // Redirect to overview if logged in successfully
        if (response.ok) {
            window.location.href = '/';
        // Show correct error message if login failed
        } else {
            const errors = await response.json();
            if (errors.errors.old_password) {
                setOldPasswordIncorrect(true);
            }
            if (errors.errors.new_password2) {
                setNewPasswordError(errors.errors.new_password2[0]);
            }
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Change Password"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Old password</span>
                    </div>
                    <input
                        name="old_password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            oldPasswordIncorrect && "input-error"
                        )}
                        value={oldPassword}
                        onInput={() => setOldPasswordIncorrect(false)}
                        onChange={(e) => setOldPassword(e.target.value)}
                    />
                </label>
                {oldPasswordIncorrect && (
                    <span className="text-error">
                        Old password incorrect
                    </span>
                )}
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">New password</span>
                    </div>
                    <input
                        name="new_password1"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword1}
                        onChange={(e) => setNewPassword1(e.target.value)}
                    />
                </label>
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Confirm new password</span>
                    </div>
                    <input
                        name="new_password2"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword2}
                        onInput={() => setNewPasswordError(false)}
                        onChange={(e) => setNewPassword2(e.target.value)}
                    />
                </label>
                {newPasswordError && (
                    <span className="text-error">
                        {newPasswordError}
                    </span>
                )}

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                >
                    Change password
                </button>
            </form>
        </div>
    );
}

export default App;
