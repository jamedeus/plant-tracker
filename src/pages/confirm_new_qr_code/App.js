import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { sendPostRequest, parseDomContext } from 'src/util';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';
import PlantDetails from 'src/components/PlantDetails.js';
import GroupDetails from 'src/components/GroupDetails.js';
import { openErrorModal } from 'src/components/ErrorModal';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faXmark, faCheck } from '@fortawesome/free-solid-svg-icons';

const Layout = ({ type, thumbnailUrl, handleConfirm }) => {
    return (
        <div className="flex flex-col gap-8 text-center my-auto">
            <p className="text-lg font-bold">
                Is this the new QR code for your {type}?
            </p>
            {thumbnailUrl && (
                <div className="mx-auto p-4 bg-base-200 rounded-3xl">
                    <img
                        className="max-h-[50vh] rounded-xl object-contain"
                        src={thumbnailUrl}
                    />
                </div>
            )}

            {/* Confirm/cancel buttons */}
            <div className="flex gap-4 mx-auto mb-8">
                <button
                    className="btn btn-error btn-square text-white"
                    onClick={() => window.location.href = "/"}
                    title="Don't change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faXmark} />
                </button>
                <button
                    className="btn btn-success btn-square text-white"
                    onClick={handleConfirm}
                    title="Change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faCheck} />
                </button>
            </div>
        </div>
    );
};

Layout.propTypes = {
    type: PropTypes.oneOf([
        'plant',
        'group'
    ]).isRequired,
    thumbnailUrl: PropTypes.string,
    handleConfirm: PropTypes.func.isRequired
};

function App() {
    // Load UUIDs and plant/group state from django template context
    const type = parseDomContext("type");
    const instance = parseDomContext("instance");
    const newUuid = parseDomContext("new_uuid");

    // Reload if user navigates to page by pressing back button (change QR code
    // request likely expired, replace with register or manage page)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                window.location.reload();
            }
        };

        // Add listener on mount, remove on unmount
        window.addEventListener('pageshow', handleBackButton);
        return () => {
            window.removeEventListener('pageshow', handleBackButton);
        };
    }, []);

    const handleConfirm = async () => {
        const response = await sendPostRequest(
            '/change_uuid',
            {
                uuid: instance.uuid,
                new_id: newUuid
            }
        );
        // Reload page if changed successfully
        if (response.ok) {
            window.location.reload();
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    };

    // Plant/group details card shown when title is clicked
    const DetailsDropdown = useMemo(() => (
        <div className="details-card">
            <div className="card-body">
                {type === 'plant' ? (
                    <PlantDetails
                        species={instance.species}
                        pot_size={instance.pot_size}
                        description={instance.description}
                    />
                ) : (
                    <GroupDetails
                        location={instance.location}
                        description={instance.description}
                    />
                )}
            </div>
        </div>
    ), []);

    return (
        <div className="container flex flex-col mx-auto h-screen">
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title={instance.display_name}
                titleOptions={DetailsDropdown}
            />
            <Layout
                type={type}
                thumbnailUrl={instance.thumbnail}
                handleConfirm={handleConfirm}
            />
        </div>
    );
}

export default App;
