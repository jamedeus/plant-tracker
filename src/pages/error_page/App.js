import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';

function App({ errorMessage }) {
    return (
        <div
            className="container flex flex-col full-screen mx-auto items-center gap-16"
            data-testid="error-page"
        >
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title="Error"
                showScanButton={false}
            />

            <div className="flex flex-col text-center text-lg mt-[15vh]">
                {errorMessage}
            </div>

            <Link to="/" className='btn btn-accent' discover="none">
                Go to Overview
            </Link>
        </div>
    );
}

export default App;

App.propTypes = {
    errorMessage: PropTypes.string.isRequired
};
