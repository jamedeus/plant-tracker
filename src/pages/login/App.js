import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [showError, setShowError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post FormData to backend
        const formData = new FormData(formRef.current);
        const response = await fetch('/accounts/login/', {
            method: 'POST',
            body: formData,
            headers: {
                'Accept': 'application/json, text/plain, */*',
                "X-CSRFToken": Cookies.get('csrftoken')
            }
        });
        // Redirect if logged in successfully
        if (response.ok) {
            // Redirect to url in querystring if present
            const params = new URL(window.location.href).searchParams;
            if (params.get('next')) {
                window.location.href = params.get('next');
            // Redirect to overview if no querystring
            } else {
                window.location.href = '/';
            }
        // Show error text if login failed
        } else {
            setShowError(true);
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Login"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Username</span>
                    </div>
                    <input
                        name="username"
                        type="text"
                        autoCapitalize="off"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={username}
                        onInput={() => setShowError(false)}
                        onChange={e => setUsername(e.target.value)}
                    />
                </label>
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">Password</span>
                    </div>
                    <input
                        name="password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={password}
                        onInput={() => setShowError(false)}
                        onChange={e => setPassword(e.target.value)}
                    />
                </label>

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                    disabled={username.length < 1 || password.length < 1}
                >
                    Login
                </button>

                {showError && (
                    <span className="text-error text-center">
                        Invalid username or password
                    </span>
                )}
            </form>
        </div>
    );
}

export default App;
