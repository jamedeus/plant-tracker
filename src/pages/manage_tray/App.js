import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { sendPostRequest, parseDomContext, localToUTC, pastTense } from 'src/util';
import EditModal from 'src/components/EditModal';
import TrayDetailsForm from 'src/forms/TrayDetailsForm';
import Navbar from 'src/components/Navbar';
import PlantCard from 'src/components/PlantCard';
import DatetimeInput from 'src/components/DatetimeInput';
import FilterColumn from 'src/components/FilterColumn';
import { useToast } from 'src/context/ToastContext';
import { useTheme } from 'src/context/ThemeContext';
import DetailsCard from 'src/components/DetailsCard';
import TrayDetails from 'src/components/TrayDetails';
import AddPlantsModal, { openAddPlantsModal } from './AddPlantsModal';
import RemovePlantsModal, { openRemovePlantsModal } from './RemovePlantsModal';
import { useErrorModal } from 'src/context/ErrorModalContext';

function App() {
    // Load context set by django template
    const [tray, setTray] = useState(() => {
        return parseDomContext("tray");
    });
    const [plantDetails, setPlantDetails] = useState(() => {
        return parseDomContext("details");
    });
    // Contains list of objects with name and uuid of every plant in database
    // DO NOT mutate (used to generate add/remove plant menu options)
    const options = parseDomContext("options");

    // Create state to track whether selecting plants from list
    const [selectingPlants, setSelectingPlants] = useState(false);

    // Track which plants are selected (after clicking manage button)
    const selectedPlants = useRef([]);

    // Track plants column open/close state between re-renders
    const plantsOpenRef = useRef(true);

    // Ref to access timestamp input used by water all/fertilize all
    const addEventAllTimeInput = useRef(null);

    // Create ref to access edit details form
    const editDetailsRef = useRef(null);

    // Get hooks to show toast message, error modal
    const { showToast } = useToast();
    const { showErrorModal } = useErrorModal();

    // Get toggle theme option from context
    const { ToggleThemeOption } = useTheme();

    const submitEditModal = async () => {
        const payload = Object.fromEntries(
            new FormData(editDetailsRef.current)
        );
        payload["tray_id"] = tray.uuid;
        console.log(payload);

        const response = await sendPostRequest('/edit_tray', payload);
        if (response.ok) {
            // Update plant state with new values from response
            const data = await response.json();
            setTray({...tray, ...data});
        } else {
            const error = await response.json();
            showErrorModal(JSON.stringify(error));
        }
    };

    // Handler for "Water All" and "Fertilize All" buttons
    const addEventAll = async (eventType) => {
        const timestamp = localToUTC(addEventAllTimeInput.current.value);
        // Post eventType, all plant UUIDs, and timestamp to backend endpoint
        await bulkAddPlantEvents(
            eventType,
            plantDetails.map(plant => plant.uuid),
            timestamp
        );
    };

    // Creates event with specified type and timestamp for every plant in
    // selectedIds (array of UUIDs)
    const bulkAddPlantEvents = async (eventType, selectedIds, timestamp) => {
        const payload = {
            plants: selectedIds,
            event_type: eventType,
            timestamp: timestamp
        };
        const response = await sendPostRequest('/bulk_add_plant_events', payload);
        if (response.ok) {
            showToast(`All plants ${pastTense(eventType)}!`, 'blue', 5000);
            const data = await response.json();
            updatePlantTimestamps(data.plants, timestamp, eventType);
        } else {
            const error = await response.json();
            showErrorModal(JSON.stringify(error));
        }
    };

    // Map eventType taken by bulk_add_plant_events to the plantDetails state
    // key that should be updated when an event is successfully created
    const eventTypeMap = {
        water: "last_watered",
        fertilize: "last_fertilized"
    };

    // Called by bulkAddPlantEvents to update plant water/fertilize timestamps
    const updatePlantTimestamps = (updatedPlants, timestamp, eventType) => {
        let newPlantDetails = [];
        plantDetails.forEach(plant => {
            // Overwrite last_watered if UUID in JSON response
            if (updatedPlants.includes(plant.uuid)) {
                plant[eventTypeMap[eventType]] = timestamp;
            }
            newPlantDetails.push(plant);
        });
        setPlantDetails(newPlantDetails);
    };

    // Buttons used to add bulk events to plants in tray
    const PlantEventButtons = ({editing, setEditing}) => {
        const addEventTimeInput = useRef(null);

        // Handler for water button (only used in this case scope)
        const water = async () => {
            const timestamp = localToUTC(addEventTimeInput.current.value);
            await bulkAddPlantEvents('water', selectedPlants.current, timestamp);
            setEditing(false);
        };

        // Handler for fertilize button (only used in this case scope)
        const fertilize = async () => {
            const timestamp = localToUTC(addEventTimeInput.current.value);
            await bulkAddPlantEvents('fertilize', selectedPlants.current, timestamp);
            setEditing(false);
        };

        switch(editing) {
            case(true):
                return (
                    <>
                        <div
                            className="flex mx-auto mb-4"
                            data-testid="addEventTimeInput"
                        >
                            <DatetimeInput inputRef={addEventTimeInput} />
                        </div>
                        <div className="flex">
                            <button
                                className="btn btn-outline mx-auto"
                                onClick={() => setEditing(false)}
                            >
                                Cancel
                            </button>
                            <button
                                className="btn btn-outline btn-info mx-auto"
                                onClick={water}
                            >
                                Water
                            </button>
                            <button
                                className="btn btn-outline btn-success mx-auto"
                                onClick={fertilize}
                            >
                                Fertilize
                            </button>
                        </div>
                    </>
                );
            case(false):
                return (
                    <div className="flex">
                        <button
                            className="btn btn-outline mx-auto"
                            onClick={() => setEditing(true)}>
                            Manage
                        </button>
                    </div>
                );
        }
    };

    PlantEventButtons.propTypes = {
        editing: PropTypes.bool,
        setEditing: PropTypes.func
    };

    return (
        <div className="container flex flex-col mx-auto mb-8">
            <Navbar
                menuOptions={
                    <>
                        <li><a onClick={() => window.location.href = "/"}>
                            Overview
                        </a></li>
                        <li><a onClick={openAddPlantsModal}>
                            Add plants
                        </a></li>
                        <li><a onClick={openRemovePlantsModal}>
                            Remove plants
                        </a></li>
                        <ToggleThemeOption />
                    </>
                }
                title={tray.display_name}
                titleOptions={
                    <DetailsCard>
                        <TrayDetails
                            location={tray.location}
                            description={tray.description}
                        />
                    </DetailsCard>
                }
            />

            <DatetimeInput inputRef={addEventAllTimeInput} />
            <div className="flex mx-auto mb-8">
                <button
                    className="btn btn-info m-2"
                    onClick={() => addEventAll('water')}
                >
                    Water All
                </button>
                <button
                    className="btn btn-success m-2"
                    onClick={() => addEventAll('fertilize')}
                >
                    Fertilize All
                </button>
            </div>

            <FilterColumn
                title="Plants"
                contents={plantDetails}
                CardComponent={PlantCard}
                editing={selectingPlants}
                selected={selectedPlants}
                openRef={plantsOpenRef}
            >
                <PlantEventButtons
                    editing={selectingPlants}
                    setEditing={setSelectingPlants}
                />
            </FilterColumn>

            <EditModal title="Edit Details" onSubmit={submitEditModal}>
                <TrayDetailsForm
                    formRef={editDetailsRef}
                    name={tray.name}
                    location={tray.location}
                    description={tray.description}
                />
            </EditModal>

            <AddPlantsModal
                trayID={tray.uuid}
                options={options}
                plantDetails={plantDetails}
                setPlantDetails={setPlantDetails}
            />

            <RemovePlantsModal
                trayID={tray.uuid}
                plantDetails={plantDetails}
                setPlantDetails={setPlantDetails}
            />
        </div>
    );
}

export default App;
