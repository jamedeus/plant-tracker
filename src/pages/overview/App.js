import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import Navbar from 'src/components/Navbar';
import { useTheme } from 'src/context/ThemeContext';
import GroupCard from 'src/components/GroupCard';
import PlantCard from 'src/components/PlantCard';
import { sendPostRequest, parseDomContext } from 'src/util';
import FilterColumn from 'src/components/FilterColumn';
import FloatingFooter from 'src/components/FloatingFooter';
import Setup from './Setup';
import PrintModal from './PrintModal';
import { useIsBreakpointActive } from "src/useBreakpoint";
import clsx from 'clsx';

const PlantsCol = ({ plants, editing, selectedPlants }) => {
    const openRef = useRef(true);

    return (
        <FilterColumn
            title="Plants"
            contents={plants}
            CardComponent={PlantCard}
            editing={editing}
            selected={selectedPlants}
            openRef={openRef}
            ignoreKeys={[
                'uuid',
                'created',
                'last_watered',
                'last_fertilized',
                'thumbnail'
            ]}
            sortByKeys={[
                {key: 'created', display: 'Added'},
                {key: 'display_name', display: 'Name'},
                {key: 'species', display: 'Species'},
                {key: 'last_watered', display: 'Watered'}
            ]}
            defaultSortKey='created'
            storageKey='overviewPlantsColumn'
        />
    );
};

PlantsCol.propTypes = {
    plants: PropTypes.array.isRequired,
    editing: PropTypes.bool.isRequired,
    selectedPlants: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

const GroupsCol = ({ groups, editing, selectedGroups }) => {
    const openRef = useRef(true);

    return (
        <FilterColumn
            title="Groups"
            contents={groups}
            CardComponent={GroupCard}
            editing={editing}
            selected={selectedGroups}
            openRef={openRef}
            ignoreKeys={[
                'uuid',
                'created'
            ]}
            sortByKeys={[
                {key: 'created', display: 'Added'},
                {key: 'name', display: 'Name'},
                {key: 'location', display: 'Location'}
            ]}
            defaultSortKey='created'
            storageKey='overviewGroupsColumn'
        />
    );
};

GroupsCol.propTypes = {
    groups: PropTypes.array.isRequired,
    editing: PropTypes.bool.isRequired,
    selectedGroups: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

// Render correct components for current state objects
const Layout = ({ plants, groups, selectedPlants, selectedGroups, editing, plantsColRef, groupsColRef, printModalRef }) => {
    // Determines if 2-column layout or single centered column
    const twoColumns = plants.length > 0 && groups.length > 0;

    return (
        <div className={clsx(
            'grid grid-cols-1 mx-auto',
            twoColumns && 'md:grid-cols-2'
        )}>
            {/* Render plants column if 1 or more plants exist */}
            {plants.length > 0 && (
                <div
                    className={clsx(
                        'scroll-mt-20',
                        twoColumns && 'md:mr-12 mb-8 md:mb-0'
                    )}
                    ref={plantsColRef}
                >
                    <PlantsCol
                        plants={plants}
                        editing={editing}
                        selectedPlants={selectedPlants}
                    />
                </div>
            )}
            {/* Render groups column if 1 or more groups exist */}
            {groups.length > 0 && (
                <div
                    className={clsx(
                        'scroll-mt-20',
                        twoColumns && 'md:ml-12'
                    )}
                    ref={groupsColRef}
                >
                    <GroupsCol
                        groups={groups}
                        editing={editing}
                        selectedGroups={selectedGroups}
                    />
                </div>
            )}
            {/* Render setup instructions if database is empty */}
            {plants.length === 0 && groups.length === 0 && (
                <Setup printModalRef={printModalRef} />
            )}
        </div>
    )
};

Layout.propTypes = {
    plants: PropTypes.array.isRequired,
    groups: PropTypes.array.isRequired,
    selectedPlants: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    selectedGroups: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    editing: PropTypes.bool.isRequired,
    plantsColRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    groupsColRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    printModalRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

// Dropdown with links to jump to plant or group columns
// Only rendered on mobile layout (both columns always visible on desktop)
const QuickNavigation = ({ plantsColRef, groupsColRef }) => {
    const jumpToPlants = () => {
        plantsColRef.current.scrollIntoView({
            behavior: "smooth",
            block: "start"
        });
        document.activeElement.blur();
    };

    const jumpToGroups = () => {
        groupsColRef.current.scrollIntoView({
            behavior: "smooth",
            block: "start"
        });
        document.activeElement.blur();
    };

    return (
        <ul
            tabIndex={0}
            className={`menu menu-md dropdown-content mt-3 z-[99] p-2 shadow
                        bg-base-300 rounded-box w-24`}
        >
            <li className="mx-auto"><a onClick={jumpToPlants}>
                Plants
            </a></li>
            <li className="mx-auto"><a onClick={jumpToGroups}>
                Groups
            </a></li>
        </ul>
    );
};

QuickNavigation.propTypes = {
    groupsColRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    printModalRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

// Top-left menu button contents
const MenuOptions = ({ archivedOverview, toggleEditing, openPrintModal }) => {
    // Get toggle theme option from context
    const { ToggleThemeOption } = useTheme();

    switch(archivedOverview) {
        case(true):
            return (
                <>
                    <li><a onClick={toggleEditing}>
                        Edit
                    </a></li>
                    <ToggleThemeOption />
                    <li><a href='/'>
                        Main overview
                    </a></li>
                </>
            );
        case(false):
            return (
                <>
                    <li><a onClick={toggleEditing}>
                        Edit
                    </a></li>
                    <li><a onClick={openPrintModal}>
                        Print QR Codes
                    </a></li>
                    <ToggleThemeOption />
                    <li><a href='/archived'>
                        Archived plants
                    </a></li>
                </>
            );
    }
};

MenuOptions.propTypes = {
    archivedOverview: PropTypes.bool.isRequired,
    toggleEditing: PropTypes.func.isRequired,
    openPrintModal: PropTypes.func.isRequired
}

function App() {
    // Load context set by django template
    const [plants, setPlants] = useState(() => {
        return parseDomContext("plants");
    });
    const [groups, setGroups] = useState(() => {
        return parseDomContext("groups");
    });

    // True if desktop layout, false if mobile
    const desktop = useIsBreakpointActive('md');
    // True if mobile layout with stacked plant and group columns
    // False if desktop layout (side by side columns) or only one column
    const stackedColumns = !desktop && plants.length > 0 && groups.length > 0;

    // Check URL to determine if viewing main overview or archive overview
    const archivedOverview = window.location.pathname === '/archived';

    // Request new state from backend if user navigates to overview by pressing
    // back button (last watered/details may be outdated if coming from manage)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                const response = await fetch('/get_overview_state');
                if (response.ok) {
                    const data = await response.json();
                    setPlants(data['plants']);
                    setGroups(data['groups']);
                } else {
                    alert('Failed to fetch current state, page may be outdated');
                }
            }
        };

        // Add listener on mount, remove on unmount
        if (!archivedOverview) {
            window.addEventListener('pageshow', handleBackButton);
            return () => {
                window.removeEventListener('pageshow', handleBackButton);
            };
        }
    }, []);

    // Create ref for modal used to generate QR codes
    const printModalRef = useRef(null);

    // State object to track edit mode (shows checkbox for each card when true)
    const [editing, setEditing] = useState(false);

    // Track which plant and group checkboxes the user has selected
    const selectedPlants = useRef([]);
    const selectedGroups = useRef([]);

    // Refs used to jump to top of plant and group columns
    const plantsColRef = useRef(null);
    const groupsColRef = useRef(null);

    // Handler for edit option in top-left dropdown
    // Toggle editing state, clear selected, remove focus (closes dropdown)
    const toggleEditing = () => {
        setEditing(!editing);
        selectedPlants.current = [];
        selectedGroups.current = [];
        document.activeElement.blur();
    };

    // Handler for delete button that appears while editing
    const handleDelete = () => {
        // Send delete request for each selected plant, remove uuid from state
        selectedPlants.current.forEach(async plant_id => {
            await sendPostRequest('/delete_plant', {plant_id: plant_id});
        });
        setPlants(plants.filter(
            plant => !selectedPlants.current.includes(plant.uuid))
        );

        // Send delete request for each selected group, remove uuid from state
        selectedGroups.current.forEach(async group_id => {
            await sendPostRequest('/delete_group', {group_id: group_id});
        });
        setGroups(groups.filter(
            group => !selectedGroups.current.includes(group.uuid))
        );

        // Reset editing state
        setEditing(false);
    };

    // Handler for archive button (main overview) and un-archive button
    // (archive overview) that appear while editing. POSTS selected plants and
    // groups to backend then removes from frontend state.
    // Takes bool argument (true if archiving, false if un-archiving)
    const handleArchive = (archived) => {
        // Send archive request for each selected plant, remove uuid from state
        selectedPlants.current.forEach(async plant_id => {
            await sendPostRequest(
                '/archive_plant',
                {plant_id: plant_id, archived: archived}
            );
        });
        setPlants(plants.filter(
            plant => !selectedPlants.current.includes(plant.uuid))
        );

        // Send archive request for each selected group, remove uuid from state
        selectedGroups.current.forEach(async group_id => {
            await sendPostRequest(
                '/archive_group',
                {group_id: group_id, archived: archived}
            );
        });
        setGroups(groups.filter(
            group => !selectedGroups.current.includes(group.uuid))
        );

        // Reset editing state
        setEditing(false);
    };

    return (
        <div className="container flex flex-col min-h-screen mx-auto pb-16">
            <Navbar
                menuOptions={<MenuOptions
                    archivedOverview={archivedOverview}
                    toggleEditing={toggleEditing}
                    openPrintModal={() => printModalRef.current.open()}
                />}
                title={archivedOverview ? "Archived" : "Plant Overview"}
                titleOptions={stackedColumns ? (
                    <QuickNavigation
                        plantsColRef={plantsColRef}
                        groupsColRef={groupsColRef}
                    />
                ) : null}
            />

            <Layout
                plants={plants}
                groups={groups}
                selectedPlants={selectedPlants}
                selectedGroups={selectedGroups}
                editing={editing}
                plantsColRef={plantsColRef}
                groupsColRef={groupsColRef}
                printModalRef={printModalRef}
            />

            <FloatingFooter visible={editing}>
                <button className="btn btn-neutral mr-4" onClick={() => setEditing(false)}>
                    Cancel
                </button>
                {archivedOverview ? (
                    <button className="btn mx-4" onClick={() => handleArchive(false)}>
                        Un-archive
                    </button>
                ) : (
                    <button className="btn mx-4" onClick={() => handleArchive(true)}>
                        Archive
                    </button>
                )}
                <button className="btn btn-error ml-4" onClick={handleDelete}>
                    Delete
                </button>
            </FloatingFooter>

            <PrintModal ref={printModalRef} />
        </div>
    );
}

export default App;
