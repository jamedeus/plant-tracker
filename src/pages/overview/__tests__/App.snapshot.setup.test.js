import renderer from 'react-test-renderer';
import createMockContext from 'src/testUtils/createMockContext';
import { ThemeProvider } from 'src/context/ThemeContext';
import { ErrorModalProvider } from 'src/context/ErrorModalContext';
import App from '../App';

describe('App', () => {
    it('matches snapshot when no models exist (setup)', () => {
        // Create mock state objects
        createMockContext('plants', []);
        createMockContext('trays', []);

        // Render App, confirm matches snapshot
        const component = renderer.create(
            <ThemeProvider>
                <ErrorModalProvider>
                    <App />
                </ErrorModalProvider>
            </ThemeProvider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
