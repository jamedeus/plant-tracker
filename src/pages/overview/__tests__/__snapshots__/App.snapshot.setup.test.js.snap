// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot when no models exist (setup) 1`] = `
<div
  className="container flex flex-col min-h-screen mx-auto pb-16"
>
  <div
    className="navbar bg-base-100 mb-4 sticky top-0 z-[99]"
  >
    <div
      className="navbar-start"
    >
      <div
        className="dropdown"
      >
        <div
          className="btn btn-ghost btn-circle"
          role="button"
          tabIndex={0}
        >
          <svg
            className="h-5 w-5"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h7"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
          </svg>
        </div>
        <ul
          className="menu menu-md dropdown-content mt-3 z-[99]
                                    p-2 shadow bg-base-300 rounded-box w-52"
          tabIndex={0}
        >
          <li>
            <a
              onClick={[Function]}
            >
              Edit
            </a>
          </li>
          <li>
            <a
              onClick={[Function]}
            >
              Print QR Codes
            </a>
          </li>
          <li>
            <a
              onClick={[Function]}
            >
              Light mode
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="navbar-center"
    >
      <a
        className="btn btn-ghost text-3xl"
      >
        Plant Overview
      </a>
    </div>
    <div
      className="navbar-end"
    />
  </div>
  <div
    className="flex flex-col mx-auto text-center my-auto px-8"
  >
    <p
      className="text-2xl"
    >
      No plants found!
    </p>
    <ul
      className="steps steps-vertical my-8"
    >
      <li
        className="step"
      >
        Print QR codes on sticker paper
      </li>
      <li
        className="step"
      >
        Add a sticker to each plant pot
      </li>
      <li
        className="step"
      >
        Scan codes to register plants!
      </li>
    </ul>
    <button
      className="btn btn-accent text-lg"
      onClick={[Function]}
    >
      Print QR Codes
    </button>
  </div>
  <dialog
    className="modal"
  >
    <div
      className="modal-box text-center flex flex-col"
    >
      <h3
        className="font-bold text-lg mb-6"
      >
        Fetching QR Codes
      </h3>
      <span
        className="loading loading-spinner loading-lg mx-auto"
      />
      <div
        className="modal-action mx-auto"
      >
        <form
          method="dialog"
        >
          <button
            className="btn"
          >
            Cancel
          </button>
        </form>
      </div>
    </div>
  </dialog>
  <dialog
    className="modal"
  >
    <div
      className="modal-box text-center flex flex-col"
    >
      <h3
        className="font-bold text-lg mb-6"
      >
        Error
      </h3>
      <p>
        The URL_PREFIX environment variable is not set, check docker config
      </p>
      <div
        className="modal-action mx-auto"
      >
        <form
          method="dialog"
        >
          <button
            className="btn"
          >
            OK
          </button>
        </form>
      </div>
    </div>
  </dialog>
</div>
`;
